gpytorch/variational/natural_variational_distribution.py

```diff
@@ -0,0 +1,169 @@
+
+    :param int num_inducing_points: Size of the variational distribution. This implies that the variational mean
+        should be this size, and the variational covariance matrix should have this many rows and columns.
+    :param batch_shape: Specifies an optional batch size
+        for the variational parameters. This is useful for example when doing additive variational inference.
+    :type batch_shape: :obj:`torch.Size`, optional
+    :param float mean_init_std: (Default: 1e-3) Standard deviation of gaussian noise to add to the mean initialization.
+
+    """
+
+    def __init__(self, num_inducing_points, batch_shape=torch.Size([]), mean_init_std=1e-3, **kwargs):
+        super().__init__(num_inducing_points=num_inducing_points, batch_shape=batch_shape, mean_init_std=mean_init_std)
+        scaled_mean_init = torch.zeros(num_inducing_points)
+        neg_prec_init = torch.eye(num_inducing_points, num_inducing_points).mul(-0.5)
+        scaled_mean_init = scaled_mean_init.repeat(*batch_shape, 1)
+        neg_prec_init = neg_prec_init.repeat(*batch_shape, 1, 1)
+
+        # eta1 and eta2 parameterization of the variational distribution
+        self.register_parameter(name="natural_vec", parameter=torch.nn.Parameter(scaled_mean_init))
+        self.register_parameter(name="natural_mat", parameter=torch.nn.Parameter(neg_prec_init))
+
+    def forward(self):
+        mean, chol_covar = _NaturalToMuVarSqrt.apply(self.natural_vec, self.natural_mat)
+        return MultivariateNormal(mean, CholLazyTensor(chol_covar))
+
+    def initialize_variational_distribution(self, prior_dist):
+        prior_prec = prior_dist.covariance_matrix.inverse()
+        prior_mean = prior_dist.mean
+        noise = torch.randn_like(prior_mean).mul_(self.mean_init_std)
+
+        self.natural_vec.data.copy_((prior_prec @ prior_mean).add_(noise))
+        self.natural_mat.data.copy_(prior_prec.mul(-0.5))
+
+
+def _triangular_inverse(A, upper=False):
+    eye = torch.eye(A.size(-1), dtype=A.dtype, device=A.device)
+    return eye.triangular_solve(A, upper=upper).solution
+
+
+def _phi_for_cholesky_(A):
+    "Modifies A to be the phi function used in differentiating through Cholesky"
+    A.tril_().diagonal(offset=0, dim1=-2, dim2=-1).mul_(0.5)
+    return A
+
+
+def _cholesky_backward(dout_dL, L, L_inverse):
+    # c.f. https://github.com/pytorch/pytorch/blob/25ba802ce4cbdeaebcad4a03cec8502f0de9b7b3/
+    #      tools/autograd/templates/Functions.cpp
+    A = L.transpose(-1, -2) @ dout_dL
+    phi = _phi_for_cholesky_(A)
+    grad_input = (L_inverse.transpose(-1, -2) @ phi) @ L_inverse
+    # Symmetrize gradient
+    return grad_input.add(grad_input.transpose(-1, -2)).mul_(0.5)
+
+
+class _NaturalToMuVarSqrt(torch.autograd.Function):
+    @staticmethod
+    def _forward(nat_mean, nat_covar):
+        try:
+            L_inv = torch.cholesky(-2.0 * nat_covar, upper=False)
+        except RuntimeError as e:
+            if str(e).startswith("cholesky"):
```

Comment:
nit: maybe more future-proof to do `str(e).lower()`?

Reply:
The Pytorch developers should have just raised a `numpy.LinAlgError` :'(